src/Lib.hs

```diff
@@ -43,76 +48,183 @@ startPetSelection user goldRemaining = do
+        , userRoster = opponentRoster
+        , userItemList = itemListEmpty
+        }
+    putStrLn "\nOPPONENT TEAM"
+    putStrLn $ display (opponentRoster)
+
+    -- BATTLE
+    battleRoster user userOpponent
+
+
+battleRoster :: User -> User -> IO ()
+battleRoster user1 user2 = do
+  -- get pets
+  printBar
+  putStrLn "NEW CHALLENGER:"
+  let r1pet = getRosterFirstWhere (userRoster user1) healthPositive
+  let r2pet = getRosterFirstWhere (userRoster user2) healthPositive
+  let user1Pet = fromJust r1pet
+  let user2Pet = fromJust r2pet
+  printPetBattle user1Pet user2Pet
+
+  -- battle pets
+  let (user1Pet', user2Pet') = battlePets user1Pet user2Pet
+  putStrLn "\nResult:"
+  printPetBattle user1Pet' user2Pet'
+
+  -- create updated rosters
+  let r1' = replacePet user1Pet user1Pet' (userRoster user1)
+  let r2' = replacePet user2Pet user2Pet' (userRoster user2)
+  
+  -- create updated users
+  let user1' = user1 { userRoster = r1' }
+  let user2' = user2 { userRoster = r2' }
+  
+  threadDelay 3000000 --sleep for a million microseconds, or one second
+
+  -- detect if either player is out of pets
+  -- TODO how to handle ties?
+  if (isRosterEmpty (userRoster user1')) == True
+    then putStrLn $ "\n**\n" ++ (userName user1') ++ " LOSES\n**"
+  else if (isRosterEmpty (userRoster user2')) == True
+    then putStrLn $ "\n**\n" ++ (userName user1') ++ " LOSES\n**"
+  else battleRoster user1' user2'
+  
+
+isRosterEmpty :: Roster -> Bool
+isRosterEmpty roster = if (isNothing (getRosterFirstWhere roster healthPositive)) then True else False
+
+
+battlePets :: Pet -> Pet -> (Pet, Pet) -- dont hate me
+battlePets p1 p2 = do
+  -- ATTACK
+  let p1HealthRemaining = getHealth (petHealthRemaining p1) - getAttack (petAttack p2)
+  let p2HealthRemaining = getHealth (petHealthRemaining p2) - getAttack (petAttack p1)
+
+  -- create new pets with new health
+  let p1' = p1 { petHealthRemaining = (Health p1HealthRemaining) }
+  let p2' = p2 { petHealthRemaining = (Health p2HealthRemaining) }
+  
+  -- check results
+  if ((getHealth (petHealthRemaining p1') < 0) || (getHealth (petHealthRemaining p2') < 0))
+    then (p1', p2')  -- one or both pets are dead
+    else battlePets p1' p2'  -- not dead battle again
+
+-- keep asking
+keepAskingWhere :: Read a => String -> (a -> Bool) -> IO a
+keepAskingWhere s p = do
+  putStr s
+  hFlush stdout
+  mresult <- readMaybe <$> getLine
+  case mresult of
+    Nothing -> keepAskingWhere s p
+    Just a -> if p a then pure a else keepAskingWhere s p
+
+
+keepAsking :: Read a => String -> IO a
+keepAsking s = keepAskingWhere s (const True)
+
+
+betweenInclusive :: Int -> Int -> Int -> Bool
+betweenInclusive a b x = a <= x && x <= b
 
 
 --
 -- ATTRIBUTES
 --
-newtype Name = Name {getName::String} deriving (Show, IsString)
-newtype Attack = Attack {getAttack::Int} deriving Show
-newtype Health = Health {getHealth::Int} deriving Show
+newtype Id = Id {getId::Int} deriving (Num, Ord, Eq)
```

Comment:
```suggestion
newtype Id = Id { getId::Int }
  deriving (Num, Ord, Eq)
```

Consider formatting these like this, there are no awards for fewest number of lines or spaces.